File: src/drivers/clamav.ts
import { spawn } from 'child_process';
import path from 'path';
import { Tail } from 'tail';
import { AntivirusDriver } from './base';
import {
  ClamAVConfig,
  EngineHealth,
  EngineInfo,
  LogEntry,
  RawEngineResult,
  UnifiedResult,
  WatchOptions,
} from '../types';
import { Logger } from '../utils/logger';

const DEFAULT_TIMEOUT = 60000;
const DEFAULT_POLL_INTERVAL = 100;

/**
 * ClamAV Driver
 *
 * Interface:
 * - RTS Log: watches log file for on-access scan results (clamonacc output)
 * - Manual Scan: executes clamdscan binary
 */
export class ClamAVDriver extends AntivirusDriver {
  private clamConfig: ClamAVConfig;

  constructor(config: ClamAVConfig, logger: Logger) {
    super(config, logger);
    this.clamConfig = config;
  }

  async rtsWatch(
    filePath: string,
    options: WatchOptions = { timeout: DEFAULT_TIMEOUT, pollInterval: DEFAULT_POLL_INTERVAL }
  ): Promise<UnifiedResult> {
    const startTime = Date.now();
    const fileId = path.basename(filePath);

    if (!this.clamConfig.rtsEnabled) {
      this.logger.info('RTS not enabled for ClamAV, skipping RTS watch');
      return this.createUnifiedResult(
        filePath,
        fileId,
        'clean',
        'rts',
        startTime,
        undefined,
        { skipped: true, reason: 'RTS not enabled' }
      );
    }

    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        tail.unwatch();
        this.logger.warn(`RTS watch timeout for ${filePath}`);
        resolve(
          this.createUnifiedResult(
            filePath,
            fileId,
            'clean',
            'rts',
            startTime,
            undefined,
            { timeout: true }
          )
        );
      }, options.timeout);

      const tail = new Tail(this.clamConfig.rtsLogPath, {
        fromBeginning: false,
        follow: true,
      });

      tail.on('line', (line: string) => {
        const entry = this.parseLogEntry(line);
        if (entry && this.pathMatches(entry.filePath, filePath)) {
          clearTimeout(timeout);
          tail.unwatch();
          resolve(
            this.createUnifiedResult(
              filePath,
              fileId,
              entry.status,
              'rts',
              startTime,
              entry.signature,
              { logEntry: line }
            )
          );
        }
      });

      tail.on('error', (error: Error) => {
        clearTimeout(timeout);
        tail.unwatch();
        this.logger.error('Tail error', { error });
        reject(error);
      });
    });
  }

  async manualScan(filePath: string): Promise<UnifiedResult> {
    const startTime = Date.now();
    const fileId = path.basename(filePath);

    const result = await this.executeScan(filePath);
    const parsed = this.parseManualScanOutput(result);

    return {
      ...parsed,
      filePath,
      fileId,
      duration: Date.now() - startTime,
    };
  }

  private executeScan(filePath: string): Promise<RawEngineResult> {
    return new Promise((resolve) => {
      // clamdscan --fdpass --stdout --no-summary <file>
      const args = ['--fdpass', '--stdout', '--no-summary', filePath];
      const proc = spawn(this.clamConfig.scanBinaryPath, args);

      let stdout = '';
      let stderr = '';

      proc.stdout.on('data', (data) => {
        stdout += data.toString();
      });

      proc.stderr.on('data', (data) => {
        stderr += data.toString();
      });

      proc.on('close', (exitCode) => {
        resolve({
          engine: 'clamav',
          exitCode: exitCode ?? -1,
          stdout,
          stderr,
        });
      });

      proc.on('error', (error) => {
        resolve({
          engine: 'clamav',
          exitCode: -1,
          stdout,
          stderr: error.message,
        });
      });
    });
  }

  protected parseLogEntry(logLine: string): LogEntry | null {
    // ClamAV on-access log format (clamonacc):
    // /path/to/file: Eicar-Test-Signature FOUND
    // /path/to/file: OK
    const foundMatch = logLine.match(/^(.+):\s+(.+)\s+FOUND$/);
    if (foundMatch) {
      return {
        timestamp: new Date(),
        filePath: foundMatch[1],
        status: 'infected',
        signature: foundMatch[2],
        raw: logLine,
      };
    }

    const okMatch = logLine.match(/^(.+):\s+OK$/);
    if (okMatch) {
      return {
        timestamp: new Date(),
        filePath: okMatch[1],
        status: 'clean',
        raw: logLine,
      };
    }

    return null;
  }

  protected parseManualScanOutput(result: RawEngineResult): UnifiedResult {
    const startTime = Date.now();
    const lines = result.stdout.trim().split('\n');

    for (const line of lines) {
      const entry = this.parseLogEntry(line);
      if (entry) {
        return this.createUnifiedResult(
          entry.filePath,
          path.basename(entry.filePath),
          entry.status,
          'manual',
          startTime,
          entry.signature,
          result
        );
      }
    }

    // Exit codes: 0 = clean, 1 = virus found, 2+ = error
    const status = result.exitCode === 0 ? 'clean' : result.exitCode === 1 ? 'infected' : 'error';

    return this.createUnifiedResult(
      '',
      '',
      status,
      'manual',
      startTime,
      undefined,
      result
    );
  }

  async checkHealth(): Promise<EngineHealth> {
    try {
      // Execute scan binary with --version
      const result = await new Promise<RawEngineResult>((resolve) => {
        const proc = spawn(this.clamConfig.scanBinaryPath, ['--version']);
        let stdout = '';
        let stderr = '';

        proc.stdout.on('data', (data) => { stdout += data.toString(); });
        proc.stderr.on('data', (data) => { stderr += data.toString(); });

        proc.on('close', (exitCode) => {
          resolve({ engine: 'clamav', exitCode: exitCode ?? -1, stdout, stderr });
        });

        proc.on('error', (error) => {
          resolve({ engine: 'clamav', exitCode: -1, stdout, stderr: error.message });
        });
      });

      const healthy = result.exitCode === 0 || result.stdout.includes('ClamAV');
      const versionMatch = result.stdout.match(/ClamAV\s+([\d.]+)/);

      return {
        engine: 'clamav',
        healthy,
        version: versionMatch?.[1],
        lastCheck: new Date(),
        error: healthy ? undefined : result.stderr,
      };
    } catch (error) {
      return {
        engine: 'clamav',
        healthy: false,
        lastCheck: new Date(),
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  getInfo(): EngineInfo {
    return {
      engine: 'clamav',
      available: true,
      rtsEnabled: this.clamConfig.rtsEnabled,
      manualScanAvailable: true,
    };
  }

  private pathMatches(logPath: string, targetPath: string): boolean {
    return path.resolve(logPath) === path.resolve(targetPath);
  }
}

File: internal/drivers/clamav.go
package drivers

import (
	"bytes"
	"context"
	"log/slog"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/nxadm/tail"
	"github.com/rophy/av-scanner/internal/cache"
	"github.com/rophy/av-scanner/internal/config"
)

// Matches: /path/to/file: Signature FOUND
var clamavFoundRegex = regexp.MustCompile(`(.+):\s+(.+)\s+FOUND$`)

type ClamAVDriver struct {
	config config.DriverConfig
	logger *slog.Logger
	cache  *cache.DetectionCache
	ctx    context.Context
	cancel context.CancelFunc
}

func NewClamAVDriver(cfg config.DriverConfig, logger *slog.Logger, detectionCache *cache.DetectionCache) *ClamAVDriver {
	ctx, cancel := context.WithCancel(context.Background())
	return &ClamAVDriver{
		config: cfg,
		logger: logger.With("driver", "clamav"),
		cache:  detectionCache,
		ctx:    ctx,
		cancel: cancel,
	}
}

// Start begins the background log watcher
func (d *ClamAVDriver) Start() error {
	if _, err := os.Stat(d.config.RTSLogPath); err != nil {
		d.logger.Warn("RTS log file not accessible, background watcher not started", "path", d.config.RTSLogPath)
		return nil
	}

	go d.watchLog()
	d.logger.Info("Background log watcher started", "path", d.config.RTSLogPath)
	return nil
}

// Stop stops the background log watcher
func (d *ClamAVDriver) Stop() {
	d.cancel()
}

func (d *ClamAVDriver) watchLog() {
	t, err := tail.TailFile(d.config.RTSLogPath, tail.Config{
		Follow:    true,
		ReOpen:    true,
		MustExist: true,
		Location:  &tail.SeekInfo{Offset: 0, Whence: os.SEEK_END},
	})
	if err != nil {
		d.logger.Error("Failed to start log tailer", "error", err)
		return
	}
	defer t.Cleanup()

	for {
		select {
		case <-d.ctx.Done():
			t.Stop()
			return
		case line := <-t.Lines:
			if line == nil || line.Err != nil {
				continue
			}
			d.processLogLine(line.Text)
		}
	}
}

func (d *ClamAVDriver) processLogLine(line string) {
	if matches := clamavFoundRegex.FindStringSubmatch(line); matches != nil {
		absPath, err := filepath.Abs(matches[1])
		if err != nil {
			absPath = matches[1]
		}

		d.cache.Add(absPath, &cache.Detection{
			FilePath:  matches[1],
			Status:    "infected",
			Signature: matches[2],
			Raw:       line,
		})
		d.logger.Debug("Cached detection", "path", absPath, "signature", matches[2])
	}
}

func (d *ClamAVDriver) Engine() config.EngineType {
	return config.EngineClamAV
}

func (d *ClamAVDriver) Config() config.DriverConfig {
	return d.config
}

func (d *ClamAVDriver) RTSWatch(filePath string, opts WatchOptions) (*ScanResult, error) {
	startTime := time.Now()
	fileID := filepath.Base(filePath)

	absPath, err := filepath.Abs(filePath)
	if err != nil {
		absPath = filePath
	}

	timeout := time.After(opts.Timeout)
	ticker := time.NewTicker(50 * time.Millisecond)
	defer ticker.Stop()

	for {
		select {
		case <-timeout:
			d.logger.Warn("RTS watch timeout", "filePath", filePath)
			return &ScanResult{
				Status:    StatusClean,
				Engine:    d.Engine(),
				Phase:     PhaseRTS,
				FilePath:  filePath,
				FileID:    fileID,
				Timestamp: time.Now(),
				Duration:  time.Since(startTime).Milliseconds(),
				Raw:       map[string]bool{"timeout": true},
			}, nil

		case <-ticker.C:
			if cached, found := d.cache.Get(absPath); found {
				status := StatusClean
				if cached.Status == "infected" {
					status = StatusInfected
				}

				return &ScanResult{
					Status:    status,
					Engine:    d.Engine(),
					Signature: cached.Signature,
					Phase:     PhaseRTS,
					FilePath:  filePath,
					FileID:    fileID,
					Timestamp: time.Now(),
					Duration:  time.Since(startTime).Milliseconds(),
					Raw:       map[string]string{"logEntry": cached.Raw},
				}, nil
			}
		}
	}
}

func (d *ClamAVDriver) ManualScan(filePath string) (*ScanResult, error) {
	startTime := time.Now()
	fileID := filepath.Base(filePath)

	// clamdscan --fdpass --stdout --no-summary <file>
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(d.config.Timeout)*time.Millisecond)
	defer cancel()

	cmd := exec.CommandContext(ctx, d.config.ScanBinaryPath, "--fdpass", "--stdout", "--no-summary", filePath)
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	exitCode := 0
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else if ctx.Err() == context.DeadlineExceeded {
			return nil, ctx.Err()
		} else {
			return nil, err
		}
	}

	output := strings.TrimSpace(stdout.String())
	d.logger.Debug("Manual scan completed", "exitCode", exitCode, "output", output)

	// Parse output for signature
	var status ScanStatus
	var signature string

	if matches := clamavFoundRegex.FindStringSubmatch(output); matches != nil {
		status = StatusInfected
		signature = matches[2]
	} else {
		// Exit codes: 0 = clean, 1 = virus found, 2+ = error
		switch exitCode {
		case 0:
			status = StatusClean
		case 1:
			status = StatusInfected
		default:
			status = StatusError
		}
	}

	return &ScanResult{
		Status:    status,
		Engine:    d.Engine(),
		Signature: signature,
		Phase:     PhaseManual,
		FilePath:  filePath,
		FileID:    fileID,
		Timestamp: time.Now(),
		Duration:  time.Since(startTime).Milliseconds(),
		Raw: map[string]interface{}{
			"exitCode": exitCode,
			"stdout":   output,
			"stderr":   stderr.String(),
		},
	}, nil
}

func (d *ClamAVDriver) CheckHealth() (*EngineHealth, error) {
	health := &EngineHealth{
		Engine:    d.Engine(),
		LastCheck: time.Now(),
	}

	if _, err := os.Stat(d.config.RTSLogPath); err != nil {
		health.Healthy = false
		health.Error = "RTS log not accessible: " + d.config.RTSLogPath
		return health, nil
	}

	health.Healthy = true
	return health, nil
}

func (d *ClamAVDriver) GetInfo() EngineInfo {
	return EngineInfo{
		Engine:              d.Engine(),
		Available:           true,
		RTSEnabled:          true,
		ManualScanAvailable: true,
	}
}
